File: binding/python/bench_test.py
# Copyright 2022 The Ip2Region Authors. All rights reserved.
# Use of this source code is governed by a Apache2.0-style
# license that can be found in the LICENSE file.

# xdb searcher bench test on 2025/10/30
# Author Leon<chenxin619315@gmail.com>

import io
import sys
import argparse
import time
import ip2region.util as util
import ip2region.searcher as xdb

def create_searcher(db_path, cache_policy):
    # open the source xdb file
    handle = io.open(db_path, "rb")

    # verify the xdb file
    # @Note: do NOT call it every time you create a searcher since this will slow
    # down the search response.
    # @see the verify function for details.
    util.verify(handle)

    # get the ip version from header
    header = util.load_header(handle)
    version = util.version_from_header(header)
    if version is None:
        handle.close()
        raise Exception("failed to get version from header")

    searcher = None
    if cache_policy == "file":
        searcher = xdb.new_with_file_only(version, db_path)
    elif cache_policy == "vectorIndex":
        v_index = util.load_vector_index(handle)
        searcher = xdb.new_with_vector_index(version, db_path, v_index)
    elif cache_policy == "content":
        c_buffer = util.load_content(handle)
        searcher = xdb.new_with_buffer(version, c_buffer)
    else:
        raise ValueError("invalid cache_policy `{}`".format(cache_policy))

    handle.close()
    return searcher


def run_bench_test(db_path: str, src_path: str, cache_policy: str):
    # create the searcher
    searcher = None
    try:
        searcher = create_searcher(args.db, args.cache_policy)
    except Exception as e:
        print("failed to create searcher: {}".format(str(e)))
        return

    print("searcher ->", searcher)
    # read the source lines and do the search test
    count, total_secs = 0, 0
    try:
        handle = io.open(src_path, "rb")
        while True:
            line = handle.readline().decode("utf-8").strip()
            # ignore empty  or comment line
            if len(line) < 1:
                break

            if line[0] == "#":
                continue

            # line splits
            ps = line.split("|", 2)
            if len(ps) != 3:
                raise Exception(f"invalid ip segment line `{line}`")

            # ip parse and compare
            start_time = time.time()
            sip_bytes = util.parse_ip(ps[0])
            eip_bytes = util.parse_ip(ps[1])
            if util.ip_compare(sip_bytes, eip_bytes) > 0:
                raise Exception(f"start ip({ps[0]}) should not be greater than end ip({ps[1]})")

            # do the search test
            for ip_bytes in [sip_bytes, eip_bytes]:
                count  = count + 1
                region = searcher.search(ip_bytes)
                if region != ps[2]:
                    raise Exception(f"failed to search({util.ip_to_string(ip_bytes)}) with {region} != {ps[2]}")
            
            total_secs = total_secs + (time.time() - start_time)

        # resource cleanup
        handle.close()
    except Exception as e:
        print(f"bench failed: {str(e)}")
        return

    # print the stats info
    each_us = total_secs * 1000_000 / count
    print(f"Bench finished, {{cachePolicy: {cache_policy}, total: {count}, took: {total_secs:.3f} s, cost: {each_us:.0f} μs/op}}");

    # resource cleanup
    searcher.close()


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        add_help=True,
        prog="python bench_test.py",
        description="ip2region bench test script",
        usage="%(prog)s [command option]"
    )

    # check the args
    parser.add_argument('--db', help='ip2region binary xdb file path')
    parser.add_argument('--src', help='source ip text file path')
    parser.add_argument('--cache-policy', help='cache policy: file/vectorIndex/content, default: vectorIndex', default="vectorIndex")
    args = parser.parse_args()
    if (args.db is None) or (args.src is None):
        parser.print_help()
        sys.exit()

    # run the search test
    run_bench_test(args.db, args.src, args.cache_policy)

File: binding/lua_c/bench_test.lua
-- Copyright 2022 The Ip2Region Authors. All rights reserved.
-- Use of this source code is governed by a Apache2.0-style
-- license that can be found in the LICENSE file.
--
-- ---
-- @Author Lion <chenxin619315@gmail.com>
-- @Date   2022/06/30

-- set the package to load the current xdb_searcher.so
package.path = "./?.lua;" .. package.path
package.cpath = "./?.so;" .. package.cpath
local xdb = require("xdb_searcher")

function printHelp()
    print("lua bench_test.lua [command options]")
    print("options: ")
    print(" --db string             ip2region binary xdb file path")
    print(" --src string            source ip text file path")
    print(" --cache-policy string   cache policy: file/vectorIndex/content")
end

if #arg < 2 then
    printHelp()
    return
end

-- parser the command line args
local dbFile, srcFile = "", ""
local cachePolicy = "vectorIndex"
for _, r in ipairs(arg) do
    if string.len(r) < 5 then
        -- continue and do nothing here
    elseif string.sub(r, 1, 2) ~= "--" then
        -- continue and do nothing here
    else
        for k, v in string.gmatch(string.sub(r, 3), "([^=]+)=([^%s]+)") do
            if k == "db" then
                dbFile = v
            elseif k == "src" then
                srcFile = v
            elseif k == "cache-policy" then
                cachePolicy = v
            else
                print(string.format("undefined option `%s`", r))
                return
            end

            -- break the match iterate
            break
        end
    end
end

-- print(string.format("dbFile=%s, srcFile=%s, cachePolicy=%s", dbFile, srcFile, cachePolicy))
if string.len(dbFile) < 2 or string.len(srcFile) < 2 then
    printHelp()
    return
end

-- verify the xdb from header
if xdb.verify(dbFile) == false then
    print(string.format("failed to verify the xdb file: %s", dbFile))
    return
end

-- detect the version from the xdb header
header, err = xdb.load_header(dbFile)
if err ~= nil then
    print(string.format("failed to load header: %s", err))
    return
end

version, err = xdb.version_from_header(header);
if err ~= nil then
    print(string.format("failed to detect version from header: %s", err))
    return
end

-- create the searcher based on the cache-policy
local searcher, v_index, content
if cachePolicy == "file" then
    searcher, err = xdb.new_with_file_only(version, dbFile)
    if err ~= nil then
        print(string.format("failed to create searcher: %s", err))
        return
    end
elseif cachePolicy == "vectorIndex" then
    v_index, err = xdb.load_vector_index(dbFile)
    if err ~= nil then
        print(string.format("failed to load vector index: %s", err))
        return
    end

    searcher, err = xdb.new_with_vector_index(version, dbFile, v_index)
    if err ~= nil then
        print(string.format("failed to create vector index searcher: %s", err))
        return
    end
elseif cachePolicy == "content" then
    content, err = xdb.load_content(dbFile)
    if err ~= nil then
        print(string.format("failed to load xdb content from '%s'", dbFile))
        return
    end

    searcher, err = xdb.new_with_buffer(version, content)
    if err ~= nil then
        print(string.format("failed to create content buffer searcher: %s", err))
        return
    end
else
    print(string.format("undefined cache-policy `%s`", cachePolicy))
    return
end

-- do the bench test
local handle = io.open(srcFile, "r")
if handle == nil then
    print(string.format("failed to open src text file `%s`", handle))
    return
end

local lines = handle:lines()
local sip_str, eip_str, s_region, region, err = "", "", "", "", 0
local count, t_time, c_time = 0, 0, 0
local s_time = xdb.now()
for l in lines do
    if string.len(l) < 1 then
        -- continue and do nothing here
    else
        for v1, v2, v3 in string.gmatch(l, "([^|]+)|([^|]+)|([^\n]+)") do
            -- print(sip_str, eip_str, region)
            sip_str = v1
            eip_str = v2
            s_region = v3
            break
        end

        t_time = xdb.now()
        sip_bytes, err = xdb.parse_ip(sip_str)
        if err ~= nil then
            print(string.format("invalid start ip `%s`", sip_str))
            return
        end

        eip_bytes, err = xdb.parse_ip(eip_str)
        if err ~= nil then
            print(string.format("invalid end ip `%s`", sip_str))
            return
        end

        if xdb.ip_compare(sip_bytes, eip_bytes) > 0 then
            print(string.format("start ip(%s) should not be greater than end ip(%s)\n", sip_str, eip_str))
            return
        end

        for _, ip_bytes in ipairs({sip_bytes, eip_bytes}) do
            region, err = searcher:search(ip_bytes)
            if err ~= nil then
                print(string.format("failed to search ip `%s`", xdb.ip_to_string(ip_bytes)))
                return
            end

            -- check the region
            if region ~= s_region then
                print(string.format("failed search(%s) with (%s != %s)\n", xdb.ip_to_string(ip_bytes), region, s_region))
                return
            end

            count = count + 1
        end

        -- increase the time costs
        c_time = c_time + xdb.now() - t_time
    end
end

-- resource cleanup
searcher:close()
if v_index ~= nil then
    v_index:close()
end
if content ~= nil then
    content:close()
end

xdb.cleanup()

-- print the stats
local avg_costs = 0
if count > 0 then
    avg_costs = c_time / count
end
print(string.format("Bench finished, {cachePolicy: %s, total: %d, took: %.3f s, cost: %.3f μs/op}",
         cachePolicy, count, (xdb.now() - s_time)/1e6, c_time / count))